File: scrape_perfect.py
# scrape_perfect.py
"""
Python script to download the Smart Qatar homepage and all referenced assets
from the Wayback Machine (id_ snapshot) and save them locally preserving the
original directory structure under the project's `public/` folder.

The script performs the following steps:
1. Fetch the raw archived HTML from the id_ snapshot URL.
2. Extract all asset URLs from `src`, `href`, `srcset` and CSS `url(...)`.
3. Convert each asset URL to a Wayback download URL.
4. Download each asset, creating necessary sub‑directories under `public/`.
5. Rewrite the HTML to reference the locally saved assets and write it to
   `public/index.html` (or appropriate language folder).

This script uses only the Python standard library (`urllib`, `re`, `os`,
`html.parser`) to avoid external dependencies.
"""
import os
import re
import urllib.request
from urllib.parse import urljoin, urlparse
from html.parser import HTMLParser

# Configuration – snapshot base URL (id_ version)
SNAPSHOT_URL = "https://web.archive.org/web/20220621003153id_/http://smart.gov.qa/en"
# Base URL of the original site for replacing asset paths
ORIGINAL_BASE = "http://smart.gov.qa"
# Local public directory where assets will be stored
PUBLIC_DIR = os.path.join(os.getcwd(), "public")

class AssetHTMLParser(HTMLParser):
    def __init__(self):
        super().__init__()
        self.assets = []
        self.html_parts = []

    def handle_starttag(self, tag, attrs):
        # Reconstruct the original tag
        attr_str = " ".join(f'{k}="{v}"' for k, v in attrs)
        self.html_parts.append(f"<{tag} {attr_str}>" if attr_str else f"<{tag}>")
        for name, value in attrs:
            if name in ("src", "href", "srcset"):
                # Split srcset if present
                for src in value.split(','):
                    url = src.strip().split(' ')[0]
                    if url:
                        self.assets.append(url)

    def handle_endtag(self, tag):
        self.html_parts.append(f"</{tag}>")

    def handle_data(self, data):
        self.html_parts.append(data)

    def handle_comment(self, data):
        self.html_parts.append(f"<!--{data}-->")

    def get_html(self):
        return "".join(self.html_parts)

def download_url(url, local_path):
    os.makedirs(os.path.dirname(local_path), exist_ok=True)
    try:
        urllib.request.urlretrieve(url, local_path)
        print(f"Downloaded {url} -> {local_path}")
    except Exception as e:
        print(f"Failed to download {url}: {e}")

def main():
    # 1. Fetch archived HTML
    with urllib.request.urlopen(SNAPSHOT_URL) as response:
        html_bytes = response.read()
    html = html_bytes.decode('utf-8', errors='ignore')

    parser = AssetHTMLParser()
    parser.feed(html)
    assets = parser.assets

    # Additionally capture CSS url(...) patterns from inline style blocks and stylesheets
    css_url_pattern = re.compile(r'url\([\'\"]?(.*?)[\'\"]?\)')
    for match in css_url_pattern.finditer(html):
        assets.append(match.group(1))

    # Normalise and dedupe asset URLs
    assets = list({a for a in assets if a})

    # 2. Process each asset
    for asset_url in assets:
        # Resolve relative URLs against the original base URL
        if asset_url.startswith('http'):
            full_original = asset_url
        else:
            # Preserve leading '/' for absolute paths
            full_original = urljoin(ORIGINAL_BASE + '/', asset_url.lstrip('/'))
        # Convert to Wayback download URL
        wayback_url = f"https://web.archive.org/web/20220621003153id_/{full_original}"
        # Determine local file path under public directory (strip leading protocol and domain)
        parsed = urlparse(full_original)
        local_path = os.path.join(PUBLIC_DIR, parsed.path.lstrip('/'))
        download_url(wayback_url, local_path)

    # 3. Rewrite HTML to point to local assets
    rewritten_html = html
    for asset_url in assets:
        if asset_url.startswith('http'):
            full_original = asset_url
        else:
            full_original = urljoin(ORIGINAL_BASE + '/', asset_url.lstrip('/'))
        parsed = urlparse(full_original)
        local_rel = parsed.path.lstrip('/')
        rewritten_html = rewritten_html.replace(asset_url, f"/{local_rel}")
        # Also replace the full original URL if present
        rewritten_html = rewritten_html.replace(full_original, f"/{local_rel}")

    # 4. Write rewritten HTML to appropriate location
    index_path = os.path.join(PUBLIC_DIR, "index.html")
    os.makedirs(os.path.dirname(index_path), exist_ok=True)
    with open(index_path, "w", encoding="utf-8") as f:
        f.write(rewritten_html)
    print(f"Rewritten HTML saved to {index_path}")

if __name__ == "__main__":
    main()

File: scrape_perfect.js
// scrape_perfect.js
// Node.js script to download the full Smart Qatar homepage from Wayback Machine (raw id_ version)
// and save all referenced assets locally preserving original paths.

const axios = require('axios');
const fs = require('fs');
const path = require('path');
const url = require('url');

// Configuration - change if you want a different snapshot
const SNAPSHOT = '20220621003153'; // yyyymmddhhmmss from Wayback URL
const BASE_ORIG = 'http://smart.gov.qa';
const WAYBACK_BASE = `http://web.archive.org/web/${SNAPSHOT}id_/${BASE_ORIG}`;

// Target directory (public folder of the project)
const PUBLIC_DIR = path.resolve(__dirname, 'public');

// Helper to ensure a directory exists
function ensureDir(filePath) {
  const dir = path.dirname(filePath);
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
}

// Download a single URL and write to localPath
async function downloadFile(remoteUrl, localPath) {
  try {
    const response = await axios.get(remoteUrl, { responseType: 'arraybuffer', timeout: 15000 });
    ensureDir(localPath);
    fs.writeFileSync(localPath, response.data);
    console.log(`✓ ${remoteUrl} → ${localPath}`);
  } catch (err) {
    console.warn(`⚠ Failed ${remoteUrl}: ${err.message}`);
  }
}

// Parse HTML for asset URLs (src, href, srcset) and CSS url(...)
function extractUrls(html) {
  const urls = new Set();
  // Simple regex for src/href attributes
  const attrRegex = /(?:src|href)\s*=\s*"([^\"]+)"/gi;
  let m;
  while ((m = attrRegex.exec(html)) !== null) {
    urls.add(m[1]);
  }
  // srcset handling (take first URL before space)
  const srcsetRegex = /srcset\s*=\s*"([^\"]+)"/gi;
  while ((m = srcsetRegex.exec(html)) !== null) {
    const parts = m[1].split(',');
    parts.forEach(p => {
      const urlPart = p.trim().split(' ')[0];
      if (urlPart) urls.add(urlPart);
    });
  }
  // CSS url(...)
  const cssUrlRegex = /url\((?:'|\")?([^'\")]+)(?:'|\")?\)/gi;
  while ((m = cssUrlRegex.exec(html)) !== null) {
    urls.add(m[1]);
  }
  return Array.from(urls);
}

(async () => {
  // 1. Download raw HTML
  const htmlUrl = `${WAYBACK_BASE}`;
  console.log('Downloading HTML from', htmlUrl);
  const htmlResp = await axios.get(htmlUrl, { timeout: 15000 });
  const rawHtml = htmlResp.data;

  // 2. Write HTML to public/index.html (or keep existing name)
  const htmlPath = path.join(PUBLIC_DIR, 'index.html');
  ensureDir(htmlPath);
  fs.writeFileSync(htmlPath, rawHtml);
  console.log('Saved HTML to', htmlPath);

  // 3. Extract asset URLs
  const assetUrls = extractUrls(rawHtml);
  console.log(`Found ${assetUrls.length} asset URLs`);

  // 4. Download each asset
  for (const assetUrl of assetUrls) {
    // Skip data URIs and external CDNs (already absolute http(s) not belonging to smart.gov.qa)
    if (assetUrl.startsWith('data:')) continue;
    let remote;
    // If URL is absolute (http/https) keep it, otherwise treat as relative to original site
    if (assetUrl.startsWith('http')) {
      remote = assetUrl;
    } else {
      // Remove leading slash for path join
      const clean = assetUrl.replace(/^\/+/,'');
      remote = `${WAYBACK_BASE}/${clean}`;
    }
    // Local path mirrors the URL path (remove leading slash)
    const localPath = path.join(PUBLIC_DIR, assetUrl.replace(/^\/+/,'') );
    await downloadFile(remote, localPath);
  }

  console.log('Download completed.');
})();
